### ISSUE TITLE
catch-all in match expression

### ISSUE THREAD

**Issue by andrewbanchich** [rust-lang/rust-clippy#12465]
### What it does

One of the biggest benefits of match expressions is they are exhaustive. However, programmers often add catch-all arms when they shouldn't.

At the time of writing the catch-all, this enum:

```rust
enum Animal {
  Monkey,
  Frog,
  Caecilian
}
```
might be matched on like:

```rust
match animal {
  Animal::Monkey => println!("it's a mammal"),
  _ => println!("it's an amphibian")
}
```

and be correct. However, since this is a catch-call, the user is opting out of the benefits of exhaustive pattern matching. If they update the definition of `Animal` to be:

```rust
enum Animal {
  Monkey,
  Frog,
  Caecilian,
  Parrot
}
```

then their program will silently start thinking that parrots are amphibians.

This lint will recommend each type be written explicitly:

```rust
match animal {
  Animal::Monkey => println!("it's a mammal"),
  Animal::Frog | Animal::Caecilian => println!("it's an amphibian")
}
```

### Advantage

Exhaustive pattern matching allows for much more fearless refactoring.

### Drawbacks

More verbose. Also, sometimes a catch-all is the most correct way to express something.

### Example

```rust
match animal {
  Animal::Monkey => println!("it's a mammal"),
  _ => println!("it's an amphibian")
}
```

Could be written as:

```rust
match animal {
  Animal::Monkey => println!("it's a mammal"),
  Animal::Frog | Animal::Caecilian => println!("it's an amphibian")
}
```

@rustbot claim

**Comment by Centri3** [comment]
We do actually already have a lint like this, `clippy::match_wildcard_for_single_variants`. We could expand it to optionally lint 2-3 variants too

**Comment by Centri3** [comment]
Oh there's actually `clippy::wildcard_enum_match_arm`. So this already exists for enums entirely. But if this would lint other types too this could still be added

**Comment by andrewbanchich** [comment]
Awesome, thanks!
